src/read/lazy.rs

```diff
@@ -0,0 +1,108 @@
+pub(crate) use imp::*;
+
+#[cfg(not(feature = "std"))]
+mod imp {
+    use alloc::sync::Arc;
+    use core::sync::atomic::{AtomicPtr, Ordering};
+    use core::{mem, ptr};
+
+    #[derive(Debug, Default)]
+    pub(crate) struct LazyArc<T> {
+        // Only written once with a value obtained from `Arc<T>::into_raw`.
+        value: AtomicPtr<T>,
+    }
+
+    impl<T> Drop for LazyArc<T> {
+        fn drop(&mut self) {
+            let value_ptr = self.value.load(Ordering::Acquire);
+            if !value_ptr.is_null() {
+                // SAFETY: all writes to `self.value` are pointers obtained from `Arc::into_raw`.
+                unsafe {
+                    Arc::from_raw(value_ptr);
```

Comment:
When using `Arc::from_raw` or `Box::from_raw` specifically just to run the `Drop` implementation, I like to wrap it in `drop()` for clarity:

```suggestion
                    drop(Arc::from_raw(value_ptr));
```